---
title: "04 NMF Analysis"
author: "Minjie Hu"
---

```{r setup, include=FALSE}
library(Seurat, quietly = T)
library(ggplot2)
library(tidyverse)
library(monocle)
library(pheatmap)
```


```{r}
  id2symbol <- read.table("data/id2symbol.txt",header=T,sep="\t",quote="\"", stringsAsFactors=F)
#Seurat V3 symbol replace "_" with "-"
  id2symbol$symbol <- gsub(pattern = "_", replacement = "-", x = id2symbol$symbol)
```

# Extract high variable gene expression matrix for NMF analysis
```{r message=FALSE}
LePin_RNAi <-readRDS("data/LePin_RNAi.RDS")
control_RNAi <-readRDS("data/control_RNAi.RDS")
symbiotic <- readRDS("data/symbiotic_seurat.RDS")
temp <-LePin_RNAi[,LePin_RNAi$Doublet == "Singlet" & LePin_RNAi$predicted.id==16]
temp <- FindVariableFeatures(temp, selection.method = "vst", nfeatures = 3000)
variable_gene <- temp@assays$RNA@var.features
temp <-control_RNAi[,control_RNAi$Doublet == "Singlet"]
temp <- FindVariableFeatures(temp, selection.method = "vst", nfeatures = 3000)
variable_gene <- unique(c(variable_gene,temp@assays$RNA@var.features))
DefaultAssay(symbiotic) <- "RNA"
exp <- symbiotic@assays$RNA@data
exp <- exp[rownames(exp) %in% variable_gene,]
exp <- as.data.frame(as.matrix(exp))
write.csv(exp, file = "NMF_symbiotic.csv")
```

# NMF analysis

  We used the run_nmf.py function from a previously published NMF analysis (Farrell et al., 2018) to analyze the data (https://github.com/YiqunW/NMF/) using the default parameters. The R codes are modified according to the Hydra scRNA-seq paper(https://github.com/cejuliano/hydra_single_cell/blob/master/SA07_NMF.Rmd).

```{r }
#import robj from python script
load("data/NMF.Robj")

#rename object
res <- result_obj
rm(result_obj)

#get list of k values that were used
getKs <- names(res)
getKs <- strsplit(getKs, "=")
getKs <- as.numeric(vapply(getKs, function(x) x[2], ""))
#this function takes as an argument an index referring to a position in the list of K values within the results and determines both the average number of genes driving each metagene and the metagene's cluster consistency score
findConsistMeta <- function(q) {
  #look at a single K value
  kval <- res[[q]]
  #initialize objects into which we will put the average number of genes/consistency scores per metagene
  consistMeta <- NULL
  goodMeta <- NULL
  #this for loop moves through each metagene
  for (j in 1:length(colnames(kval[[1]][['G']]))) {
    #get metagenes from the first replicate
    runRep <- kval[[1]]
    #first we'll identify the cells that are clustered into that metagene
    #get Cell scores
    cellScores <- runRep[['C']]
    #normalize scores so that the top cell has a score of one
    cellScores <- apply(cellScores, 1, function(x) x/max(x))
    #assign cluster ID based on top metagene score
    MaxClust <- apply(cellScores, 1, function(x) colnames(cellScores)[which.max(x)])
    #look at all cells belonging to the current metagene being considered
    clustCells <- names(MaxClust[MaxClust==colnames(cellScores)[j]])
    #next we'll count the number of genes within one order of magnitude of the top scoring gene
    #pull gene scores
    geneScores <- runRep[['G']]
    #determine how many genes are within an order of magnitude of the top gene
    geneCount <- geneScores[,paste0("X",colnames(cellScores)[j])]
    geneCount <- length(geneCount[geneCount >= (max(geneCount)/10)])
    #look at these clustered cells in the other replicates
    #initialize the objects into which we will put the results for each run (within one K, within one metagene)
    metaConsistency <- NULL
    metaGeneCount <- geneCount
    #disregard the for loop if there are no cells assigned to the cluster in question
    if (length(clustCells) != 0) {
      #this for loop moves through each replicate (excluding the first, which we already looked at)
      for (i in 2:length(kval)) {
        #get cell scores
        otherRep <- kval[[i]][['C']]
        #normalize so that the top gene has a score of 1
        otherRep <- apply(otherRep, 1, function(x) x/max(x))
        #assign cells to a cluster
        otherRep <- apply(otherRep, 1, function(x) colnames(otherRep)[which.max(x)])
        #isolate the cells that were grouped together into a cluster in the first replicate
        otherRep <- otherRep[names(otherRep) %in% clustCells]
        #ask how many of those cells belong to the same cluster
        otherRep <- table(otherRep)
        #add it to the rep list
        metaConsistency <- c(metaConsistency, max(otherRep)/length(clustCells))
        #pull the metagene name of the best hit (determined by cell overlap) and determine how many genes are within 
        #an order of magnitude of the top gene
        #pull the name of the metagene that best matches the query metagene
        otherRepMetaHit <- names(otherRep[order(-otherRep)])[1]
        #get gene scores for the rep in question
        otherRep.genes <- kval[[i]][['G']]
        otherRep.genes <- otherRep.genes[,paste0("X",otherRepMetaHit)]
        #how many genes are within an order of magnitude of the top gene?
        otherRep.genes <- length(otherRep.genes[otherRep.genes >= (max(otherRep.genes)/10)])
        #add results to the rep list
        metaGeneCount <- c(metaGeneCount, otherRep.genes)
      }
    } else {
      metaConsistency <- 0
      metaGeneCount <- 0
    }
    #compile the results per metagene
    consistMeta <- c(consistMeta,mean(metaConsistency))
    goodMeta <- c(goodMeta,mean(metaGeneCount))
  }
  #compile results into a data frame for each K value
  goodMetaCount <- data.frame(metaGeneID = colnames(kval[[1]][['G']]), consistency_score = consistMeta, gene_count = goodMeta)
  return(goodMetaCount)
}
#because running findConsistMeta can be time-consuming, we can use a parallel for loop
library(doParallel)
cl <- makeCluster(6)
registerDoParallel(cl)
#run findConsistMeta across all K values
allKconsistMeta <- foreach(i=1:length(res)) %dopar% findConsistMeta(i)
#Exclude those metagenes that don't meet QC thresholds
allKconsistMeta.IDs <- lapply(allKconsistMeta,function(x) as.character(x[,1][x[,2] > 0.6 & x[,3] > 10]))
#give each DF it's correct name
names(allKconsistMeta.IDs) <- getKs
#get number of informative metagenes for each K value
numMetaKeep <- vapply(allKconsistMeta.IDs,length,numeric(1))
numMetaKeep <- data.frame(kval = getKs, MK = numMetaKeep)
```

# Post-NMF Analysis

   We select optimal K according to the method used in the hydra scRNA-seq paper. Besides, we found when using a large K value, some metagene will result in very simialr genes. Taking both into consideration, we finally lands on K=45.

```{r Initial Metagene Filtering}
#specify the K value we want to use
x <- "K=45"
#isolate the results for desired K value
kval <- res[[x]]
#pull the results for the first replicate
runRep <- kval[[1]]
#get the cell scores
cellScores <- runRep[['C']]
#get the list of good metagene indexes for that kvalue
metaIndex <- allKconsistMeta.IDs[[gsub("K=","",x)]]
#because the gene number per metagene is an average of all replicates, there may be some metagenes in this particular 
#replicate that are on the "good" list, even if the gene in this replicate does not pass our threshold
#This just catches those exceptions and excludes them
#get gene scores
geneScores <- runRep[['G']]
#find metagenes that don't pass the gene number threshold
extraFilter <- apply(geneScores, 2, function(x) x[x > (max(x)/10)])
list.condition <- sapply(extraFilter, function(x) length(x) > 10)
extraFilter  <- names(extraFilter[list.condition])
#exclude metagenes from this replicate that don't pass the threshold
metaIndex <- metaIndex[metaIndex %in% extraFilter]
#convert to an index (instead of a column name)
metaIndex <- as.numeric(gsub("X","",metaIndex))
metaIndex <- metaIndex + 1
#get the index of "bad" metagenes in case they need to be review them later
badMetaIndex <- 1:length(rownames(cellScores))
badMetaIndex <- badMetaIndex[!(badMetaIndex %in% metaIndex)]
#exclude bad metagenes
goodCellScores <- cellScores[metaIndex,]
goodGeneScores <- geneScores[,metaIndex]
#exclude good metagenes
badCellScores <- cellScores[badMetaIndex,]
badGeneScores <- geneScores[,badMetaIndex]
```

# Filter Metagene and export top genes for each metagene
```{r}
symbiotic_mono <- readRDS("data/RNAi_symbiotic_mono.RDS")

# Compare the top 20% active cells for each metagene
sample_distribution <- function(x) {
#  temp <- colnames(goodCellScores[x])
  table(pData(symbiotic_mono)[x,]$sample)
}
goodCellScores_bi <- apply(goodCellScores, 1, function(x) x> quantile(x,0.8))
temp <- apply(goodCellScores_bi, 2, sample_distribution)
temp <- t(temp)
temp <- as.data.frame(temp)
temp$Ratio <-temp$control_RNAi/temp$LePin_RNAi

temp$meta_gene <- rownames(temp)
#read in the top 30 gene tables generated by the python script
top30 <- runRep$top30genes
top30 <- top30[,colnames(top30) %in% c(paste0("Module.",temp$meta_gene),paste0("Weights.",temp$meta_gene))]
#filter metagene with low weight genes only
temp$max_weight <- as.vector(unlist(t(top30[,colnames(top30) %in% paste0("Weights.",temp$meta_gene)][1,])))
temp <- temp[temp$max_weight>0.8,]
#rename metagenes
old2new_name <- data.frame(old=sort(as.numeric(rownames(temp))),new=1:24)
tableS1 <- top30[,(colnames(top30) %in% c(paste0("Module.",temp$meta_gene),paste0("Weights.",temp$meta_gene)))]
colnames(tableS1) <- paste0(rep(c("metagene ","weight "),24),rep(1:24,each=2))
write.table(tableS1,"figures/Supplementary Table S1.xls",sep="\t",row.names = F)
old2new_name$old <- as.factor(old2new_name$old)
temp <- left_join(temp, old2new_name,by=c("meta_gene"="old"))
#temp$new <- paste0("metagene",temp$new)
temp$new <- factor(temp$new,levels=temp$new)
metagene_summary <- temp
```


# metagene heatmap along pseudotime
```{r}
#normalized metagene activity for each cell
max_index <- apply(goodCellScores,1,max)
goodCellScores_nom <- goodCellScores/max_index
goodCellScores_nom <- goodCellScores_nom[as.character(metagene_summary$meta_gene),]
temp <- pData(symbiotic_mono)
temp <- temp[order(temp$Pseudotime),]
goodCellScores_nom <- goodCellScores_nom[,rownames(temp)]
rownames(goodCellScores_nom) <- metagene_summary$new
#pdf("figures/Metagene_pseudotime_heatmap.pdf",height=3.5,width = 6.8)
pheatmap(goodCellScores_nom,show_colnames = F,cluster_rows = F,cluster_cols = F)
#dev.off()
pheatmap(goodCellScores_nom,show_colnames = F,cluster_rows =T,cluster_cols = F)
```


```{r Metagene Plots, fig.width=10, fig.height=10, fig.cap="The first nine metagenes"}
#Create cellscores matrix
t.goodCellScores <- as.data.frame(t(goodCellScores))
t.goodCellScores <- t.goodCellScores[,metagene_summary$meta_gene]
colnames(t.goodCellScores) <- paste0("GEP",metagene_summary$new)
t.goodCellScores$orig.ident <- rownames(t.goodCellScores)

#add cell scores as metagene columns to seurat object
pData(symbiotic_mono) <- pData(symbiotic_mono)[,1:40]
pData(symbiotic_mono)$orig.ident <- rownames(pData(symbiotic_mono))
pData(symbiotic_mono) <- left_join(pData(symbiotic_mono),t.goodCellScores,by="orig.ident")
rownames(pData(symbiotic_mono)) <- pData(symbiotic_mono)$orig.ident

#export score for each cell
write.table(pData(symbiotic_mono)[,39:64],"Table S3.xls",sep="\t",quote=F)
```

# Metagene1,4,22 activity along pseudotime
```{r}
 ggplot(pData(symbiotic_mono), aes(x = Pseudotime, y = GEP1,color=sample))+geom_point()  +stat_summary_bin(fun.data = "mean_cl_boot", geom = "crossbar", binwidth = 1)+theme_classic()
 ggplot(pData(symbiotic_mono), aes(x = Pseudotime, y = GEP4,color=sample))+geom_point()  +stat_summary_bin(fun.data = "mean_cl_boot", geom = "crossbar", binwidth = 1)+theme_classic()
 ggplot(pData(symbiotic_mono), aes(x = Pseudotime, y = GEP22,color=sample))+geom_point()  +stat_summary_bin(fun.data = "mean_cl_boot", geom = "crossbar", binwidth = 1)+theme_classic()
 saveRDS(symbiotic_mono,"data/RNAi_symbiotic_mono.RDS")
```


# Software versions
This document was computed on `r format( Sys.time(), "%a %b %d %X %Y" )` with the following R package versions.
```{r session_summary, echo=FALSE, include=TRUE, comment=NA}
    sessionInfo()
```
